pallets/axelar-router/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+	let encoded_axelar_contract = Contract {
+		constructor: None,
+		functions: BTreeMap::<String, Vec<Function>>::from([(
+			AXELAR_FUNCTION_NAME.into(),
+			vec![Function {
+				name: AXELAR_FUNCTION_NAME.into(),
+				inputs: vec![
+					Param {
+						name: AXELAR_DESTINATION_CHAIN_PARAM.into(),
+						kind: ParamType::String,
+						internal_type: None,
+					},
+					Param {
+						name: AXELAR_DESTINATION_CONTRACT_ADDRESS_PARAM.into(),
+						kind: ParamType::String,
+						internal_type: None,
+					},
+					Param {
+						name: AXELAR_PAYLOAD_PARAM.into(),
+						kind: ParamType::Bytes,
+						internal_type: None,
+					},
+				],
+				outputs: vec![],
+				constant: Some(false),
+				state_mutability: Default::default(),
+			}],
+		)]),
+		events: Default::default(),
+		errors: Default::default(),
+		receive: false,
+		fallback: false,
+	}
+	.function(AXELAR_FUNCTION_NAME)
+	.map_err(|_| "cannot retrieve Axelar contract function")?
+	.encode_input(&[
+		Token::String(
+			String::from_utf8(target_chain).map_err(|_| "target chain conversion error")?,
+		),
+		// Ensure that the target contract is correctly converted to hex.
+		//
+		// The `to_string` method on the H160 is returning a string containing an ellipsis, such
+		// as: 0x1234…7890
+		Token::String(format!("0x{}", hex::encode(target_contract.0))),
+		Token::Bytes(serialized_msg),
+	])
+	.map_err(|_| "cannot encode input for Axelar contract function")?;
+
+	Ok(encoded_axelar_contract)
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// The origin that is allowed to set the gateway address we accept
+		/// messages from
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// The target of the messages comming from other chains
+		type Gateway: MessageReceiver<Middleware = Self::Middleware, Origin = DomainAddress>;
+
+		/// Middleware used by the gateway
+		type Middleware: From<AxelarId>;
+
+		/// The target of the messages comming from this chain
+		type Transactor: EthereumTransactor;
+
+		/// Checker to ensure an evm account code is registered
+		type EvmAccountCodeChecker: PreConditions<(H160, H256), Result = bool>;
+	}
+
+	#[pallet::storage]
+	pub type Configuration<T: Config> = StorageMap<_, Twox64Concat, ChainName, AxelarConfig>;
+
+	#[pallet::storage]
+	pub type ChainNameById<T: Config> = StorageMap<_, Twox64Concat, AxelarId, ChainName>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		ConfigSet {
+			name: ChainName,
+			config: AxelarConfig,
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Domain not found.
+		DomainNotFound,
+
+		/// Router not found.
+		RouterNotFound,
+
+		/// Emit when the evm account code is not registered
+		ContractCodeNotMatch,
```

Comment:
Nit: To align with other error naming scheme

```suggestion
		ContractCodeMismatch,
```